From 0d019fbad1525a147c7a37ca3c86fa5d8c0ef644 Mon Sep 17 00:00:00 2001
From: Frank Li <Frank.Li@freescale.com>
Date: Thu, 19 Apr 2012 12:31:15 +0800
Subject: [PATCH 314/435] ENGR00180222 Add dump clock tree to dot

Please use below command to convert to png
dot -Tpng -O d.txt

Signed-off-by: Frank Li <Frank.Li@freescale.com>
---
 test/mxc_clock_test/Makefile           |  3 +-
 test/mxc_clock_test/dump-clocks-dot.sh | 50 ++++++++++++++++++++++++++++++++++
 2 files changed, 52 insertions(+), 1 deletion(-)
 create mode 100755 test/mxc_clock_test/dump-clocks-dot.sh

diff --git a/test/mxc_clock_test/Makefile b/test/mxc_clock_test/Makefile
index 806dcb1..cf0ec33 100644
--- a/test/mxc_clock_test/Makefile
+++ b/test/mxc_clock_test/Makefile
@@ -2,7 +2,8 @@
 EXCLUDE_LIST:=
 
 ifeq (,$(findstring $(PLATFORM), $(EXCLUDE_LIST)))
-OBJS = $(OBJDIR)/dump-clocks.sh
+OBJS = $(OBJDIR)/dump-clocks.sh \
+       $(OBJDIR)/dump-clocks-dot.sh
 else
 OBJS =
 endif
diff --git a/test/mxc_clock_test/dump-clocks-dot.sh b/test/mxc_clock_test/dump-clocks-dot.sh
new file mode 100755
index 0000000..b325947
--- /dev/null
+++ b/test/mxc_clock_test/dump-clocks-dot.sh
@@ -0,0 +1,50 @@
+#!/bin/bash
+
+if ! mount|grep -sq '/sys/kernel/debug'; then
+	mount -t debugfs none /sys/kernel/debug
+fi
+
+saved_path=$PWD
+
+#printf "%-24s %-24s %3s %9s\n" "clock" "parent" "use" "rate"
+printf "digraph clocktree {\n"
+printf "rankdir = LR;\n"
+printf "node [shape=record];\n"
+
+for foo in $(find /sys/kernel/debug/clock -type d); do
+    if [ "$foo" = '/sys/kernel/debug/clock' ]; then
+        continue
+    fi
+
+    cd $foo
+
+    use="$(cat usecount)"
+    rate="$(cat rate)"
+
+    clk="$(basename $foo)"
+    cd ..
+    parent="$(basename $PWD)"
+
+    if [ "$use" = 0 ]; then
+    	bgcolor="red"
+    else
+    	bgcolor="green"
+    fi
+
+    printf "%s [label=\"%s|rate %d|use $use\" color=%s];\n" "$clk" "$clk" "$rate" $bgcolor
+
+    if [ "$parent" = 'clock' ]; then
+        parent="virtual_Root"
+# 	printf "%s d];\n"  "$clk" "$clk" "$rate"
+    else
+    	printf "%s -> %s\n" "$parent" "$clk"
+    fi
+
+#    printf "%s [label=%s_%d];\n" "$clk" "$rate"
+#    printf "%s -> %s ;\n" "$parent" "$clk"
+# "$use" "$rate"
+
+    cd $saved_path
+done
+
+printf "}\n"
-- 
1.8.0

